File: raceEngine/src/engine/race.js
// Le deroulement d'une course : la grille, le depart, les tours, l'arrivee.

import { parkPosition, remainingDistance } from './geometry.js';
import { getInitialKartSpeed } from './stats.js';
import { awardRacePoints, getLeader } from './standings.js';
import { regenItemDecay } from './items.js';

// Le coup d'envoi. Chaque kart tire son depart : turbo pour la grande
// majorite, depart normal, ou cale d'une seconde.
function launchKarts(cfg, state, rng, now, events) {
    const race = cfg.race;

    for (const kart of state.karts) {
        kart.state = 'running';

        const roll = rng();
        if (roll < race.startTurboChance) {
            kart.boostEndTime = now + race.turboBoostMs;
            kart.absoluteVelocity = kart.stats.topSpeed;
            kart.momentum = 1;
            events.push({ type: 'startBoost', kartId: kart.id, kind: 'turbo' });
        } else if (roll < race.startTurboChance + race.startNormalChance) {
            kart.absoluteVelocity = getInitialKartSpeed(rng, kart.stats);
            events.push({ type: 'startBoost', kartId: kart.id, kind: 'normal' });
        } else {
            kart.startStallUntil = now + race.failStallMs;
            kart.absoluteVelocity = 0;
            kart.momentum = 0;
            events.push({ type: 'startBoost', kartId: kart.id, kind: 'fail' });
        }
    }
}

// Ce qu'il reste a parcourir au kart le plus en retard encore en course.
function lastRemaining(state) {
    let most = -Infinity;
    for (const kart of state.karts) {
        if (kart.finished) continue;
        const left = remainingDistance(kart);
        if (left > most) most = left;
    }
    return most;
}

function setSign(state, group, frame, now, duration) {
    state.sign = { group: group, frame: frame, until: now + duration };
}

function countdownDuration(race) {
    return race.countdownHoldMs + 2 * race.lightIntervalMs;
}

function updateRace(cfg, state, rng, now, deltaTime, events) {
    const race = cfg.race;
    const leader = getLeader(state);

    // Tour du premier. `lapCount` compte les franchissements ; le premier
    // cloture le trajet depuis la grille et non un tour, d'ou le plancher a 1.
    //
    // Suivi hors de la machine a phases, et c'est essentiel : la camera passe
    // en approche deux tours avant la fin, si bien que tenir ce compteur dans
    // la seule phase 'racing' le figeait a 3 pour les deux derniers tours de
    // chaque course. Avec lui les decotes d'objets, qui ne se resorbaient
    // alors plus jamais — une bleue lancee au troisieme tour restait a demi
    // probabilite jusqu'a l'arrivee.
    if (state.phase === 'racing' || state.phase === 'finishing') {
        const lap = Math.min(race.laps, Math.max(1, leader.lapCount));
        if (lap !== state.leaderLap) {
            state.leaderLap = lap;
            regenItemDecay(cfg, state);
        }

        // Le panneau du dernier tour. Il sort quand le PREMIER approche la
        // ligne qui ouvre ce tour -- a `flagDistance`, comme le drapeau -- et
        // reste en main jusqu'a ce que le DERNIER l'ait passee d'autant : chaque
        // kart passe devant lui en entamant son dernier tour. Borne en distance
        // et non en duree : un kart retarde a l'approche aurait vu un delai fixe
        // expirer avant son passage.
        //
        // Le dernier est relu a chaque pas, pas fige a la sortie du panneau :
        // un depassement en queue de peloton change qui ferme la marche. Et si
        // le premier prend un tour au dernier, il arrive au drapeau avant que
        // celui-ci n'entame son dernier tour : le drapeau remplace le panneau
        // (setSign), qui ne revient plus.
        //
        // Hors de la machine a phases pour la meme raison que le compteur
        // ci-dessus : a cet instant la course est deja en 'finishing' (la camera
        // s'approche deux tours avant la fin). Tenu dans la seule phase
        // 'racing', il ne sortait jamais.
        const width = cfg.world.width;
        if (!state.finalSignShown && race.laps > 1 &&
            remainingDistance(leader) - width <= race.flagDistance) {
            state.finalSignShown = true;
            setSign(state, 'laps', 'final', now, race.maxRaceMs);
        } else if (state.sign && state.sign.group === 'laps' &&
                   lastRemaining(state) - width < -race.flagDistance) {
            state.sign = null;
        }

        // Et pour CHAQUE kart, s'il est dans sa propre zone de dernier tour.
        // Le panneau du service est unique, mais la camera est propre a chaque
        // spectateur : quand le premier prend un tour au dernier, le drapeau
        // (pour lui) et le dernier tour (pour le dernier) se chevauchent, et
        // c'est au client de montrer celui du kart qu'il suit.
        for (const kart of state.karts) {
            const toLine = remainingDistance(kart) - width;
            kart.finalLapSign = state.finalSignShown && !kart.finished &&
                toLine <= race.flagDistance && toLine >= -race.flagDistance;
        }
    }

    if (state.phase === 'countdown') {
        // Un feu par seconde jusqu'au depart : les quatre images de
        // lakitu/start sont montrees dans l'ordre.
        const remaining = state.startAt - now;

        // Premiere image tenue le temps de l'attente, puis un feu par
        // intervalle. La quatrieme, le feu vert, n'apparait qu'au GO.
        const elapsed = state.countdownMs - remaining;
        const step = elapsed < race.countdownHoldMs
            ? 1
            : Math.min(3, 2 + Math.floor((elapsed - race.countdownHoldMs) / race.lightIntervalMs));
        if (!state.sign || state.sign.group !== 'start' || state.sign.frame !== step) {
            setSign(state, 'start', step, now, remaining + race.goSignMs);
        }

        if (now >= state.startAt) {
            state.phase = 'racing';
            // La quatrieme image est le feu vert : elle n'apparait qu'au
            // coup d'envoi, pas pendant le decompte.
            setSign(state, 'start', 4, now, race.goSignMs);
            launchKarts(cfg, state, rng, now, events);
            events.push({ type: 'raceStart' });
        }
        return;
    }

    if (state.phase === 'racing') {
        if (remainingDistance(leader) <= race.cameraApproachDistance) {
            state.phase = 'finishing';
            state.cameraTarget = parkPosition(cfg, race.parkFinishOffset);

            // Pas de drapeau ici : la camera se gare deux tours avant la
            // fin et la ligne reste a l'ecran tout ce temps, sortir Lakitu
            // des maintenant le laisserait plante la une demi-course. C'est
            // la phase 'finishing' qui le sort a l'approche reelle.
            events.push({ type: 'raceFinishing' });
        }
        return;
    }

    if (state.phase === 'finishing') {
        // Le drapeau ne sort qu'a l'approche reelle de la ligne, pas des le
        // repositionnement de la camera. Une fois sorti il reste en main :
        // il accompagne chaque passage, pas seulement le premier.
        if (!state.flagShown && leader && remainingDistance(leader) <= race.flagDistance) {
            state.flagShown = true;
            setSign(state, 'finish', 1, now, race.maxRaceMs);
        }

        // Deux facons de clore la course : le quota d'arrivees est atteint,
        // ou le delai large est depasse — un kart bloque ne doit pas figer
        // le service. Dans les deux cas les retardataires sont classes dans
        // l'ordre ou ils roulent.
        // Borne sur le plateau reel : avec moins de karts que prevu (des
        // personnages retires du tirage, `roster`), le quota fixe ne serait
        // jamais atteint et chaque course irait jusqu'a `maxRaceMs`. A 8 karts,
        // rien ne change.
        const quota = Math.min(race.stopAtFinisher, Math.max(1, state.karts.length - 1));
        const quotaReached = state.finishOrder.length >= quota;
        const timedOut = now > state.startAt + race.maxRaceMs;

        if ((quotaReached || timedOut) && !state.resultsAt) {
            const stragglers = state.karts.filter(kart => !kart.finished)
                .sort((a, b) => a.rank - b.rank);
            for (const kart of stragglers) {
                kart.finished = true;
                kart.finishRank = state.finishOrder.length + 1;
                state.finishOrder.push(kart.id);
            }

            awardRacePoints(cfg, state);

            // La derniere course du bloc porte le classement general : on
            // laisse le temps de le lire.
            const isFinalRace = state.gpRound >= cfg.grandPrix.races;
            state.resultsAt = now + (isFinalRace ? race.finalResultsDelayMs : race.resultsDelayMs);
            state.phase = 'results';
            events.push({ type: 'raceFinished' });
        }
        return;
    }

    if (state.phase === 'results' && now >= state.resultsAt) {
        // Le service en tire une course neuve : c'est lui qui detient
        // createWorldState et les connexions a prevenir. `gpComplete` lui
        // dit s'il repart sur un bloc neuf — scores remis a zero et grille
        // tiree au sort — ou sur la course suivante du bloc en cours.
        events.push({
            type: 'raceOver',
            order: state.finishOrder.slice(),
            gpRound: state.gpRound,
            gpComplete: state.gpRound >= cfg.grandPrix.races,
            gpPoints: Object.assign({}, state.gpPoints)
        });
        state.resultsAt = now + race.resultsDelayMs;
    }
}

export {
    countdownDuration,
    updateRace,
};

File: raceEngineCpp/src/engine/race.cpp
#include "engine/race.hpp"

#include <algorithm>
#include <limits>
#include <cmath>

#include "engine/geometry.hpp"
#include "engine/standings.hpp"
#include "engine/step.hpp"

namespace engine {

namespace {

void set_sign(WorldState& state, const char* group, int frame,
              double now, double duration) {
    state.signGroup = group;
    state.signFrame = frame;
    state.signUntil = now + duration;
}

// Le coup d'envoi. Chaque kart tire SON depart : turbo pour la grande majorite,
// depart normal, ou cale d'une seconde.
void launch_karts(const config::Config& cfg, WorldState& state, Rng& rng,
                  double now, std::vector<Event>& events) {
    const config::RaceCfg& race = cfg.race;

    for (Kart& kart : state.karts) {
        kart.state = KartState::Running;

        const double roll = rng.next();
        Event ev;
        ev.type = EventType::StartBoost;
        ev.kartId = kart.id;

        if (roll < race.startTurboChance) {
            kart.boostEndTime = now + race.turboBoostMs;
            kart.absoluteVelocity = kart.stats->topSpeed;
            kart.momentum = 1;
            ev.value = 0; // turbo
        } else if (roll < race.startTurboChance + race.startNormalChance) {
            kart.absoluteVelocity = kart.stats->topSpeed * rng.range(0.85, 0.95);
            ev.value = 1; // normal
        } else {
            // Depart rate : le kart reste sur place, moteur noye.
            kart.startStallUntil = now + race.failStallMs;
            kart.absoluteVelocity = 0;
            kart.momentum = 0;
            ev.value = 2; // fail
        }
        events.push_back(ev);
    }
}

} // namespace

double countdown_duration(const config::RaceCfg& race) {
    return race.countdownHoldMs + 2 * race.lightIntervalMs;
}

void update_race(const config::Config& cfg, WorldState& state, Rng& rng,
                 double now, double deltaTime, std::vector<Event>& events) {
    (void)deltaTime;
    const config::RaceCfg& race = cfg.race;
    const Kart* leader = get_leader(state);
    if (!leader) return;

    // Tour du premier. `lapCount` compte les FRANCHISSEMENTS ; le premier
    // cloture le trajet depuis la grille et non un tour, d'ou le plancher a 1.
    //
    // Suivi HORS de la machine a phases, et c'est essentiel : la camera passe en
    // approche deux tours avant la fin, si bien que tenir ce compteur dans la
    // seule phase 'racing' le figeait pour les deux derniers tours de chaque
    // course.
    if (state.phase == Phase::Racing || state.phase == Phase::Finishing) {
        const int lap = std::min(race.laps, std::max(1, leader->lapCount));
        if (lap != state.leaderLap) state.leaderLap = lap;

        // Le panneau du dernier tour. Il sort quand le PREMIER approche la ligne
        // qui ouvre ce tour -- a `flagDistance`, comme le drapeau -- et reste en
        // main jusqu'a ce que le DERNIER l'ait passee d'autant. Le dernier est
        // relu a chaque pas (un depassement en queue change qui ferme la
        // marche) ; si le premier prend un tour au dernier, le drapeau le
        // remplace avant. Borne en distance et non en duree, hors de la machine
        // a phases : voir race.js, qui porte le raisonnement complet.
        const double width = cfg.world.width;
        const double leaderToLine = leader->finishDistance - leader->totalDistance - width;
        if (!state.finalSignShown && race.laps > 1 && leaderToLine <= race.flagDistance) {
            state.finalSignShown = true;
            set_sign(state, "laps", 0, now, race.maxRaceMs);
        } else if (state.signGroup == "laps") {
            double lastRemaining = -std::numeric_limits<double>::infinity();
            for (const Kart& kart : state.karts) {
                if (kart.finished) continue;
                lastRemaining = std::max(lastRemaining, kart.finishDistance - kart.totalDistance);
            }
            if (lastRemaining - width < -race.flagDistance) {
                state.signGroup.clear();
                state.signFrame = 0;
            }
        }

        // Et pour CHAQUE kart, s'il est dans sa propre zone de dernier tour : le
        // client montre le panneau du kart qu'il suit quand drapeau et dernier
        // tour se chevauchent (voir race.js).
        for (Kart& kart : state.karts) {
            const double toLine = kart.finishDistance - kart.totalDistance - width;
            kart.finalLapSign = state.finalSignShown && !kart.finished &&
                toLine <= race.flagDistance && toLine >= -race.flagDistance;
        }
    }

    // Le panneau s'efface tout seul.
    if (!state.signGroup.empty() && now > state.signUntil) {
        state.signGroup.clear();
        state.signFrame = 0;
    }

    if (state.phase == Phase::Countdown) {
        const double remaining = state.startAt - now;

        // Un feu par intervalle : la premiere image est tenue le temps de
        // l'attente, la quatrieme — le feu vert — n'apparait qu'au GO.
        const double elapsed = state.countdownMs - remaining;
        const int step = (elapsed < race.countdownHoldMs)
            ? 1
            : std::min(3, 2 + static_cast<int>(
                  std::floor((elapsed - race.countdownHoldMs) / race.lightIntervalMs)));

        if (state.signGroup != "start" || state.signFrame != step) {
            set_sign(state, "start", step, now, remaining + race.goSignMs);
        }

        if (now >= state.startAt) {
            state.phase = Phase::Racing;
            set_sign(state, "start", 4, now, race.goSignMs);
            launch_karts(cfg, state, rng, now, events);
            events.push_back({ EventType::RaceStart });
        }
        return;
    }

    if (state.phase == Phase::Racing) {
        const double remaining = leader->finishDistance - leader->totalDistance;

        if (remaining <= race.cameraApproachDistance) {
            state.phase = Phase::Finishing;
            state.hasCameraTarget = true;
            state.cameraTarget = park_position(cfg, race.parkFinishOffset);
            // Pas de drapeau ici : la camera se gare deux tours avant la fin et
            // la ligne reste a l'ecran tout ce temps. C'est la phase
            // 'finishing' qui sort Lakitu a l'approche REELLE.
            events.push_back({ EventType::RaceFinishing });
        }
        return;
    }

    if (state.phase == Phase::Finishing) {
        const double remaining = leader->finishDistance - leader->totalDistance;

        // Une fois sorti, le drapeau reste en main : il accompagne CHAQUE
        // passage, pas seulement le premier.
        if (!state.flagShown && remaining <= race.flagDistance) {
            state.flagShown = true;
            set_sign(state, "finish", 0, now, race.maxRaceMs);
        }

        // Deux facons de clore : le quota d'arrivees est atteint, ou le delai
        // large est depasse — un kart bloque ne doit pas figer le service. Dans
        // les deux cas les retardataires sont classes dans l'ordre ou ils
        // roulent.
        //
        // Quota borne sur le plateau reel, comme en JS : avec moins de karts
        // que prevu, le quota fixe ne serait jamais atteint.
        const int quota = std::min(race.stopAtFinisher,
                                   std::max(1, static_cast<int>(state.karts.size()) - 1));
        const bool quotaReached =
            static_cast<int>(state.finishOrder.size()) >= quota;
        const bool timedOut = now > state.startAt + race.maxRaceMs;

        if ((quotaReached || timedOut) && state.resultsAt == 0) {
            std::vector<Kart*> stragglers;
            for (Kart& kart : state.karts) {
                if (!kart.finished) stragglers.push_back(&kart);
            }
            std::sort(stragglers.begin(), stragglers.end(),
                      [](const Kart* a, const Kart* b) { return a->rank < b->rank; });

            for (Kart* kart : stragglers) {
                kart->finished = true;
                kart->finishRank = static_cast<int>(state.finishOrder.size()) + 1;
                state.finishOrder.push_back(kart->id);
            }

            award_race_points(cfg, state);

            // La derniere manche du bloc porte le classement general : on laisse
            // le temps de le lire.
            const bool isFinalRace = state.gpRound >= cfg.grandPrix.races;
            state.resultsAt = now + (isFinalRace ? race.finalResultsDelayMs
                                                 : race.resultsDelayMs);
            state.phase = Phase::Results;
            events.push_back({ EventType::RaceFinished });
        }
        return;
    }

    if (state.phase == Phase::Results && state.resultsAt > 0 && now >= state.resultsAt) {
        // Le SERVICE en tire une course neuve : c'est lui qui detient
        // `create_world_state` et les connexions a prevenir.
        Event ev;
        ev.type = EventType::RaceOver;
        ev.value = (state.gpRound >= cfg.grandPrix.races) ? 1 : 0;
        events.push_back(ev);

        // Repousse : sans ca l'evenement partirait a chaque tick jusqu'a ce que
        // le service ait fini de rebatir le monde.
        state.resultsAt = now + race.resultsDelayMs;
    }
}

} // namespace engine
